File: Problem Set 1/Credit/credit.c
#include <cs50.h>
#include <stdio.h>


bool luhn_check(long number);
void identify_card(long number);


int main(void)
{


   // ask user for input number of credit card
   long number = get_long("Number: ");


   // checking validity
   if (luhn_check(number))
   {
       identify_card(number);
   }
   else
   {
       printf("INVALID\n");
   }
   return 0;
}


bool luhn_check(long number)
{
   int sum = 0;
   int digit_count = 0;


   while (number > 0)
   {
       int digit = number % 10;


       // If the digit is in an odd position (from the end), add it directly
       if (digit_count % 2 == 0)
       {
           sum += digit;
       }
       else
       {
           // If the digit is in an even position, double it and add the digits of the product
           int product = digit * 2;
           sum += (product / 10) + (product % 10);
       }


       number /= 10;
       digit_count++;
   }


   // A valid card number has a sum divisible by 10
   return sum % 10 == 0;
}


void identify_card(long number)
{
   int length = 0;
   long start_digits = number;


   // Determine the length of the number and extract the starting digits
   while (start_digits >= 100)
   {
       start_digits /= 10;
       length++;
   }
   length += 2; // Add the last two digits


   // Check for card type
   if ((start_digits == 34 || start_digits == 37) && length == 15)
   {
       printf("AMEX\n");
   }
   else if ((start_digits >= 51 && start_digits <= 55) && length == 16)
   {
       printf("MASTERCARD\n");
   }
   else if ((start_digits / 10 == 4) && (length == 13 || length == 16))
   {
       printf("VISA\n");
   }
   else
   {
       printf("INVALID\n");
   }
}

File: Problem Set 6/credit.py
import sys


def main():
    credit_card = card_number()
    validate_card(credit_card)


def card_number():
    while True:
        card_number = input("Number: ")
        if card_number.isdigit():
            return card_number


def validate_card(credit_card):
    if len(credit_card) < 13 or len(credit_card) > 16:
        print("INVALID")
        sys.exit(0)

    even, odd = 0, 0
    reverse_digits = credit_card[::-1]

    for i, digit in enumerate(reverse_digits):
        number = int(digit)
        if i % 2 == 1:
            multiple = number * 2
            even += (multiple // 10) + (multiple % 10)
        else:
            odd += number

    checksum = (even + odd) % 10

    if checksum == 0:
        first_digit = int(credit_card[0])
        second_digit = int(credit_card[1]) if len(credit_card) > 1 else 0

        if first_digit == 3 and (second_digit == 4 or second_digit == 7):
            print("AMEX")
        elif first_digit == 5 and 1 <= second_digit <= 5:
            print("MASTERCARD")
        elif first_digit == 4:
            print("VISA")
        else:
            print("INVALID")
    else:
        print("INVALID")


if __name__ == "__main__":
    main()
